### PR TITLE
AsyncEvaluator

### LINKED ISSUE #1271: Improve evaluation during training
There are several problems with the current evaluation during training:

1. It blocks the training. For tasks with long episodes, it can significantly slow down training.
2. It only uses a single environment for evaluation. If we want to evaluate multiple episodes. It can take even more time.

### PR DESCRIPTION
**Pull request by emailweixu**
Fixes #1271 

The AsyncEvaluator creates a different process to evaluate the performance asynchronously so that it will not block the training. It also supports evaluation using parallel environments by setting TrainerConfig.num_eval_environments to number greater than one.

### CODE REVIEW

File: alf/metrics/metric.py [hunk 1]
```diff
@@ -58,14 +60,18 @@ def result(self):
         raise NotImplementedError(
             'Metrics must define a result() member function')
 
-    def gen_summaries(self, train_step=None, step_metrics=()):
+    def gen_summaries(self,
+                      train_step=None,
+                      step_metrics=(),
+                      other_steps: Dict[str, Number] = dict()):
```
Comment (hnyu) [inline alf/metrics/metric.py:66]: Can this `other_steps` just be merged with `step_metrics`?

File: alf/metrics/metric.py [hunk 2]
```diff
@@ -58,14 +60,18 @@ def result(self):
...
         """Generates summaries against train_step and all step_metrics.
 
         Args:
             train_step: (Optional) Step counter for training iterations. If None, no
                 metric is generated against the global step.
             step_metrics: (Optional) Iterable of step metrics to generate summaries
                 against.
+            other_steps: A dictionary of steps to generate summaries against.
```
Comment (hnyu) [inline alf/metrics/metric.py:73]: It's better to clarify that `other_steps` should directly contain integers instead of metrics, to differentiate from `step_metrics`.

File: alf/trainers/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            eval_dir, flush_secs=config.summaries_flush_secs)
+        alf.summary.enable_summary()
+        logging.info("Evaluator started")
+        with alf.summary.push_summary_writer(summary_writer):
+            while True:
+                job = job_queue.get()
+                if job.type == "eval":
+                    logging.info("Start evaluation")
+                    alf.summary.set_global_counter(job.global_counter)
+                    policy_trainer.Trainer._trainer_progress.update(
```
Comment (hnyu) [inline alf/trainers/evaluator.py]: Why do we need to update trainer progress here?

File: alf/trainers/evaluator.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+                        job.global_counter, job.env_steps)
+                    algorithm.load_state_dict(job.state_dict)
+                    done_queue.put(None)
+                    metrics = evaluate(env, algorithm,
+                                       config.num_eval_episodes)
+                    common.log_metrics(metrics)
+                    for metric in metrics:
+                        metric.gen_summaries(
+                            train_step=alf.summary.get_global_counter(),
```
Comment (hnyu) [inline alf/trainers/evaluator.py]: Should this just be `job.global_counter`?

File: alf/trainers/evaluator.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+        Args:
+            algorithm: the training algorithm
+            current_env_steps: current total number of environment steps.
+        """
+        job = EvalJob(
+            type="eval",
+            env_steps=int(current_env_steps),
+            global_counter=int(alf.summary.get_global_counter()),
+            state_dict=algorithm.state_dict())
```
Comment (hnyu) [inline alf/trainers/evaluator.py]: This limited number of EvalJob fields seems to make `gen_summaries` with other metrics not useful. Might have a field that store a list of steps?

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/metrics/metric.py (hunk 2 above)
Comment (emailweixu) [reply alf/metrics/metric.py:73]: The type of other_steps indicates so.

File: alf/trainers/evaluator.py (hunk 5 above)
Comment (emailweixu) [reply alf/trainers/evaluator.py]: Changed env_steps to step_metric_values.

File: alf/trainers/evaluator.py (hunk 3 above)
Comment (emailweixu) [reply alf/trainers/evaluator.py]: comment added.

File: alf/trainers/evaluator.py (hunk 4 above)
Comment (emailweixu) [reply alf/trainers/evaluator.py]: They are same because the set_global_counter above.

**Comment by breakds** [comment]
Code change LGTM. The amortized GPU memory consumption should only slightly increase if my understanding is correct.

**Review by breakds — DISMISSED** [review]

File: alf/trainers/evaluator.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+                the evaluation summaries against. Note that it needs to contain
+                "EnvironmentSteps" at least.
+        """
+        job = EvalJob(
+            type="eval",
+            step_metrics=step_metric_values,
+            global_counter=int(alf.summary.get_global_counter()),
+            state_dict=algorithm.state_dict())
+        self._job_queue.put(job)
+        self._done_queue.get()
```
Comment (le-horizon) [inline alf/trainers/evaluator.py]: When a previous eval job is still running and eval is called again, the worker is still busy working on the previous job and cannot yet take the job off the queue, so in this case, the evaluator will block on this done_queue.get()?

Maybe let's add a timer here to log to tensorboard how long this done_queue.get() is taking, so that we'll know that the eval process is blocking the training.

**Review by le-horizon — COMMENTED** [review]
Nice change, Wei!

**Review by hnyu — DISMISSED** [review]

File: alf/algorithms/config.py [hunk 7]
```diff
@@ -165,6 +166,8 @@ def __init__(self,
             eval_uncertainty (bool): whether to evluate uncertainty after training.
             num_eval_episodes (int) : number of episodes for one evaluation.
             num_eval_environments: the number of environments for evaluation.
+            async_eval: whether to do evaluation asynchronously in a different
+                process. Note that this more use more memory.
```
Comment (breakds) [inline alf/algorithms/config.py]: s/more/may/

**Review by breakds — DISMISSED** [review]

**Review by breakds — APPROVED** [review]

File: alf/trainers/evaluator.py [hunk 8]
```diff
@@ -38,29 +40,31 @@
...
+        self._async = config.async_eval
+        num_envs = config.num_eval_environments
+        seed = config.random_seed
+        if self._async:
+            ctx = mp.get_context('spawn')
+            self._job_queue = ctx.Queue()
+            self._done_queue = ctx.Queue()
+            self._worker = ctx.Process(
+                target=_worker,
+                args=(self._job_queue, self._done_queue, conf_file, num_envs,
```
Comment (hnyu) [inline alf/trainers/evaluator.py:67]: Can we directly pass `config` to `_worker` instead of letting it parse `conf_file`?

Comment (emailweixu) [reply alf/trainers/evaluator.py:67]: The subprocess is created using spawn method. Different from fork, config needs to be parsed again in the spawned subprocess anyway.

**Review by hnyu — APPROVED** [review]

File: alf/trainers/evaluator.py (hunk 6 above)
Comment (emailweixu) [reply alf/trainers/evaluator.py]: added
